### PR TITLE
Add DNAT to packetfabric_cloud_router_bgp_session and packetfabric_cloud_router_connection

### PR DESCRIPTION
**Pull request by Favouroked**
## Description

Update **nat** and add **dnat** to bgp and cloud router connection endpoints.

1. Update `packetfabric_cloud_router_bgp_session` terraform **data-source** and **resource**.
2. Add **maybe_dnat** to `packetfabric_cloud_router_connection_aws`, `packetfabric_cloud_router_connection_azure`, `packetfabric_cloud_router_connection_google`, `packetfabric_cloud_router_connection_ibm`, `packetfabric_cloud_router_connection_oracle`, `packetfabric_cloud_router_connection_port` resources.
3. Did some internal refactoring on how to handle BGP update and delete operations better
4. Added status check on the Cloud Router Connection status when updating the BGP session

## Cross-reference

PSE-10085

## Checklist

- [x] tested locally
- [ ] added/updated acceptance tests
- [x] add examples (in example folder)
- [x] generate docs

[PSE-1234]: https://packetfabric.atlassian.net/browse/PSE-1234?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ
[PSE-10085]: https://packetfabric.atlassian.net/browse/PSE-10085?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

### CODE REVIEW

**Comment by rjouhann** [comment]
@Favouroked  Could you also update the BGP session resource `packetfabric_cloud_router_bgp_session` (`resource_bgp_session.go`)? Were you able to test it in dev/locally?

**Review by rjouhann — CHANGES_REQUESTED** [review]
@Favouroked once #302 is merged (also touching the same bgp data-source), let's revisit this PR and make sure you test it locally?

**Comment by rjouhann** [comment]
@Favouroked I made some changes to your PR to add some of the missing pieces.  @medzin if you could also provide any guidance if you spot the problem, this would be great!

However, I must have missed something as it looks like I am not sending the proper payload, see error message:

```
packetfabric_cloud_router_bgp_session.crbs_3: Creating...
╷
│ Error: Status: 400, body: {"message": "nat: pre_nat_sources and pool_prefixes must be set when nat_type is overload.", "request_id": "e36fbe74-48dc-4aaa-be03-f7d8bf91579e"}
│ 
│   with packetfabric_cloud_router_bgp_session.crbs_3,
│   on main.tf line 711, in resource "packetfabric_cloud_router_bgp_session" "crbs_3":
│  711: resource "packetfabric_cloud_router_bgp_session" "crbs_3" {
│ 
```

Using TF:
```
resource "packetfabric_cloud_router_bgp_session" "crbs_3" {
  provider      = packetfabric
  circuit_id    = "PF-L3-CUST-2839140"
  connection_id = "PF-L3-CON-2910010"
  address_family = var.pf_crbs_af
  multihop_ttl   = var.pf_crbs_mhttl
  remote_asn     = var.vpn_side_asn3
  orlonger       = var.pf_crbs_orlonger
  remote_address = var.vpn_remote_address # On-Prem side
  l3_address     = var.vpn_l3_address     # PF side
  nat {                                   # example source NAT
    direction       = "output"            # or input
    nat_type        = "overload"
    pre_nat_sources = ["10.1.1.0/24", "10.1.2.0/24"]
    pool_prefixes   = ["192.168.1.50/32", "192.168.1.51/32"]
  }
  prefixes {
    prefix = "0.0.0.0/0"
    type   = "out" # Allowed Prefixes to Cloud
    order  = 0
  }
  prefixes {
    prefix = "0.0.0.0/0"
    type   = "in" # Allowed Prefixes from Cloud
    order  = 0
  }
}
```

File: internal/provider/resource_bgp_session.go [hunk 1]
```diff
@@ -394,6 +437,9 @@ func extractBgpSession(d *schema.ResourceData) packetfabric.BgpSession {
 	if md5, ok := d.GetOk("md5"); ok {
 		bgpSession.Md5 = md5.(string)
 	}
+	for _, nat := range d.Get("nat").(*schema.Set).List() {
```
Comment (medzin) [inline internal/provider/resource_bgp_session.go:440]: Does this code work properly when `nat` is not present? I would expect it to fail as `Get` will return `nil`. It is safer to use `GetOk` (so the loop will be guarded by `if`).

**Review by medzin — CHANGES_REQUESTED** [review]

File: internal/provider/resource_bgp_session.go (hunk 1 above)
Comment (rjouhann) [reply internal/provider/resource_bgp_session.go]: 1. added check `if nat, ok := d.GetOk("nat"); ok {`
2. also added `,omitempty` in struct definition for `Nat` 

However, it looks like I am still sending the `nat{}` :(

```
{"md5": "", "l3_address": "169.254.51.2/29", "address_family": "v4", "remote_address": "169.254.51.1/29", "remote_asn": 64534, "multihop_ttl": 1, "orlonger": true, "prefixes": [{"prefix": "0.0.0.0/0", "type": "in"}, {"prefix": "0.0.0.0/0", "type": "out"}], "nat": {}}
```

Comment (rjouhann) [reply internal/provider/resource_bgp_session.go]: I got it working using a pointer as suggested.

**Review by rjouhann — DISMISSED** [review]

**Review by medzin — DISMISSED** [review]

**Comment by Favouroked** [comment]
LGTM

**Review by rjouhann — APPROVED** [review]

**Pull request merged**